### PR TITLE
ACC-353: Migrations for concurrent multi-version support

### PR DESCRIPTION
**Pull request by vierbergenlars**
- [ACC-353] Add layer to manage splitting migration into multiple files
- [ACC-353] Add generic statement visitor for visitors that handle all types of Statement
- [ACC-353] Consolidate MigrationSplitter to also be a StatementCollector
- [ACC-353] Add context to StatementCollectorFactory
- [ACC-353] Avoid creating files for migrations that are completely empty
- [ACC-353] Add visitor that projects SQL change operations to a virtual database schema
- [ACC-353] Load parent migrations when testing database migrations
- [ACC-353] Add StatementCollector that generates a compatibility schema before performing the new migrations
- [ACC-353] Move compatibility statements generator to a separate class, for reuse in the AfterCompatibilitySchemaStatementCollector
- [ACC-353] Add AfterCompatibilitySchemaStatementCollector
- [ACC-353] Handle created columns in schema migration
- [ACC-353] Move migration tests to a parameterized test to eliminate code duplication
- [ACC-353] Derive before migrations from the after migrations in the tests
- [ACC-353] Split rewrite for current version from rewrite of previous version to current version
- [ACC-353] Don't use implicit `ObjectRef#toString()` to convert to an object reference
- [ACC-353] Add write support for schema compatibility dquote> dquote> Add rules, trigger functions and triggers to handle write operations to older schema versions

[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ACC-353]: https://xenitsupport.jira.com/browse/ACC-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/AfterCompatibilitySchemaStatementCollector.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.CreateViewStatement;
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.dml.SelectStatement;
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.dml.SelectStatement.SelectColumn;
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.ref.TableRef;
+import eu.xenit.contentcloud.scribe.generator.database.sql.Statement;
+import eu.xenit.contentcloud.scribe.generator.database.sql.virtual.SchemaDescriptor;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+public class AfterCompatibilitySchemaStatementCollector extends AbstractCompatibilitySchemaStatementCollector {
```
Comment (tgeens) [inline scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/AfterCompatibilitySchemaStatementCollector.java:13]: A one-line (javadoc) comment that summarizes the responsibility of each of those collectors would help a lot.

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/BeforeCompatibilitySchemaStatementCollector.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    @Override
+    protected Stream<? extends Statement> createStatementsForTableModification(
+            TableModificationDescriptor tableModificationDescriptor) {
+        return Stream.concat(
+                super.createStatementsForTableModification(tableModificationDescriptor),
+                createOneOffTriggers(tableModificationDescriptor)
+        );
+    }
+
+    private Stream<? extends Statement> createOneOffTriggers(TableModificationDescriptor table) {
```
Comment (tgeens) [inline scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/BeforeCompatibilitySchemaStatementCollector.java:59]: There are 4 ways to return from this very long method. Extracting each in a different sub-method would be nice

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/BeforeCompatibilitySchemaStatementCollector.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.ref.TriggerFunctionRef;
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.value.SqlValue;
+import eu.xenit.contentcloud.scribe.generator.database.sql.Statement;
+import eu.xenit.contentcloud.scribe.generator.database.sql.virtual.SchemaDescriptor;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import lombok.NonNull;
+
+public class BeforeCompatibilitySchemaStatementCollector extends AbstractCompatibilitySchemaStatementCollector {
```
Comment (tgeens) [inline scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/BeforeCompatibilitySchemaStatementCollector.java:26]: Add comment: this is about stopping to update/insert data into tables/columns that are about to get deleted

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/AfterCompatibilitySchemaStatementCollector.java (hunk 1 above)
Comment (tgeens) [reply scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/AfterCompatibilitySchemaStatementCollector.java:13]: My understanding: this is about creating a plain pass-through view for the current version to the default schema

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/CompatibilityStatementsGenerator.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                                        )
+                                        .build())
+                                .build()
+                ),
+                createRewritesForDeletedColumns(table, viewName)
+        );
+    }
+
+    private Stream<Statement> createRewritesForDeletedColumns(TableModificationDescriptor table, TableRef viewName) {
+        // We only need these insert and update triggers when postgres can no cleanly map back the view from the SELECT statement.
```
Comment (tgeens) [inline scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/CompatibilityStatementsGenerator.java:82]:
```suggestion
        // We only need these insert and update triggers when postgres can not cleanly map back the view from the SELECT statement.
```

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/sql/dml/DmlStatement.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.dml;
+
+import eu.xenit.contentcloud.scribe.generator.database.migrations.collector.compatibility.sql.plpgsql.PlPgSqlStatement;
+
+public interface DmlStatement extends PlPgSqlStatement {
```
Comment (tgeens) [inline scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/sql/dml/DmlStatement.java:5]: Maybe not that important in our pg-is-our-only-db context, but that feels somewhat wrong. 
Should anything from `.sql.dml` depend on vendor specific `.sql.plpgsql` ?

Comment (tgeens) [reply scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/sql/dml/DmlStatement.java:5]: Well, we probably generate pg-specific sql in the `.sql.dml` package anyway ;-)

**Review by tgeens — COMMENTED** [review]
Minor feedback - two additional questions:
* can we turn it off with a setting or request parameter ?
* how should we test that the generated versions are actually functional and compatible ? (maybe out of scope for this project, but we should have an answer for that somewhere ?)

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/sql/dml/DmlStatement.java (hunk 5 above)
Comment (vierbergenlars) [reply scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/sql/dml/DmlStatement.java:5]: >  Should anything from .sql.dml depend on vendor specific .sql.plpgsql ?

`PlPgSqlStatement` is a statement that can be used inside the context of a plpgsql language function.
All DML statements, and some additional ones (setting variables) can be used inside a function

File: scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/BeforeCompatibilitySchemaStatementCollector.java (hunk 3 above)
Comment (vierbergenlars) [reply scribe-generator/src/main/java/eu/xenit/contentcloud/scribe/generator/database/migrations/collector/compatibility/BeforeCompatibilitySchemaStatementCollector.java:26]: Yes

**Comment by vierbergenlars** [comment]
Answering your additional questions:

> can we turn it off with a setting or request parameter ?

I already talked about application-model feature flags. This would be one use-case for it.
It is unsafe (as in: will make your migrations fail or will break concurrently running versions) to toggle this feature based on a request parameter.

To initialize this feature, the first compatibility migration must be special-cased so that does not contain a operations that rewrite the previous schema views. They will fail, because no postgres schema exists. Some parts (rules and triggers) are only set up once for every view.
You won't find any special-case code in this PR, because when starting from the first data model version, it just happens to work out correctly because the "previous version" is an empty data model.

> how should we test that the generated versions are actually functional and compatible 

There are two additional "stages" to test here:

1. Do our migrations apply correctly on postgres?
    That involves only starting a database and throwing our generated SQL at it to see if it succeeds.
    You can also test the compatibility views/triggers there by doing some SELECT/UPDATE/INSERT/DELETE in the different schemas.
2. Does our application work with the generated schemas?
    To test that, we need to generate and build and execute the source code generated by Scribe, and make calls to the REST API to verify functionality.
    These tests could also verify the generated OpenAPI spec, by generating a client from the OpenAPI spec, and using that to call the REST API.
    This is a testcase that is completely independent of the underlying technology stack and only tests our public APIs. We could completely change how our database schema looks and switch to generating a Go application, and these tests would still pass.

**Review by tgeens — APPROVED** [review]

**Pull request merged**